test/server/upload-test.js

```diff
@@ -249,6 +252,135 @@ describe('Upload', function () {
         });
     });
 
+    describe('Upload storage', function () {
+        var SRC_PATH = 'test/server/data/upload/';
+
+        var tmpDirObj = null;
+
+        before(function () {
+            tmpDirObj = tmp.dirSync();
+        });
+
+        after(function () {
+            tmpDirObj.removeCallback();
+        });
+
+        function getStoredFilePath (fileName) {
+            return path.resolve(path.join(tmpDirObj.name, fileName));
+        }
+
+        function getSrcFilePath (fileName) {
+            return path.resolve(path.join(SRC_PATH, fileName));
+        }
+
+        function assertFileContentsIdentical (path1, path2) {
+            expect(fs.readFileSync(path1).toString()).eql(fs.readFileSync(path2).toString());
+        }
+
+        function assertFileExistence (filePath) {
+            expect(fs.existsSync(filePath)).to.be.true;
+        }
+
+        function assertCorrectErr (errMsg, filePath) {
+            expect(errMsg.err.indexOf('ENOENT')).not.eql(-1);
+            expect(errMsg.path).eql(filePath);
+        }
+
+        function assetCorrectFileInfo (fileInfo, fileName, filePath) {
+            expect(fileInfo.data).eql(fs.readFileSync(filePath).toString('base64'));
+            expect(fileInfo.info.name).eql(fileName);
+            expect(fileInfo.info.type).eql(mime.lookup(filePath));
+            expect(fileInfo.info.lastModifiedDate).eql(fs.statSync(filePath).mtime);
+        }
+
+        it('Should store and get file', function (done) {
+            var storage = new UploadStorage(tmpDirObj.name);
+
+            var srcFilePath    = getSrcFilePath('file-to-upload.txt');
+            var storedFilePath = getStoredFilePath('file-to-upload.txt');
+
+            storage
```

Comment:
Here and after: promise allows you avoid using such nested constructs (it's actually their puprose):
```js
storage
    .store(['file-to-upload.txt'], [filePath])
    .then(function (result) {
        expect(result.length).eql(1);
        assertCorrectErr(result[0], storedFilePath);

        return storage.get(['file-to-upload.txt']);
    })
    .then(function (result) {
        expect(result.length).eql(1);
        assertCorrectErr(result[0], storedFilePath);
        done();
    })
    .catch(done);
```